server/src/test/java/io/crate/expression/scalar/ArrayAppendFunctionTest.java

```diff
@@ -0,0 +1,205 @@
+        assertNormalize("array_append([], 1)", isLiteral(List.of(1)));
+    }
+
+    @Test
+    public void testDifferentConvertableInnerTypesLong() throws Exception {
+        assertNormalize("array_append([1::integer], 2::long)", isLiteral(Arrays.asList(1L, 2L)));
+    }
+
+    @Test
+    public void testCastTextToIntegerInnerTypes() throws Exception {
+        assertNormalize("array_append([1,2], '3')", isLiteral(Arrays.asList(1, 2, 3)));
+    }
+
+    @Test
+    public void testCastTextToDoubleInnerTypes() throws Exception {
+        assertNormalize("array_append([1,2.0], '3.0')", isLiteral(Arrays.asList(1.0, 2.0, 3.0)));
+    }
+
+    @Test
+    public void testDifferentConvertableInnerTypesDouble() throws Exception {
+        assertNormalize("array_append([1::integer], 2::double)", isLiteral(Arrays.asList(1.0, 2.0)));
+    }
+
+    @Test
+    public void test2DArrayArgumentWithArrayObject() throws Exception {
+        assertNormalize("array_append([[1,2],[3,4]], [5,6])", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(3,4), Arrays.asList(5,6))));
+    }
+
+    @Test
+    public void test2DArrayNullArgumentWithArrayObject() throws Exception {
+        assertNormalize("array_append(null, [5,6])", isLiteral(List.of(Arrays.asList(5, 6))));
+    }
+
+    @Test
+    public void test2DArrayArgumentSingleNullValueWithArrayObject() throws Exception {
+        assertNormalize("array_append([null], [5,6])", isLiteral(Arrays.asList(null, Arrays.asList(5, 6))));
+    }
+
+    @Test
+    public void test2DArrayArgumentWithNullObject() throws Exception {
+        assertNormalize("array_append([[1,2],[3,4]], null)", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(3,4), null)));
+    }
+
+    @Test
+    public void test2DArrayArgumentIncludingNullArrayWithArrayObject() throws Exception {
+        assertNormalize("array_append([[1,2],null], [5,6])", isLiteral(Arrays.asList(Arrays.asList(1,2), null, Arrays.asList(5,6))));
+    }
+
+    @Test
+    public void test2DArrayArgumentIncludingNullValueWithArrayObject() throws Exception {
+        assertNormalize("array_append([[1,2],[null, 4]], [5,6])", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(null,4), Arrays.asList(5,6))));
+    }
+
+    @Test
+    public void test2DArrayArgumentWithArrayObjectIncludingNullValue() throws Exception {
+        assertNormalize("array_append([[1,2],[3,4]], [null, 6])", isLiteral(Arrays.asList(Arrays.asList(1,2), Arrays.asList(3,4), Arrays.asList(null,6))));
+    }
+
+    @Test
+    public void testEmptyArrayNullObject() throws Exception {
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("The array argument of the `array_append` function can be of " +
+                                        "undefined type or of undefined inner type, but not when object argument " +
+                                        "is also undefined");
+        assertNormalize("array_append([], null)", null);
```

Comment:
What is the reason to not allow this?

Reply:
Appending a null value to an empty array indeed makes sense.
I tried to follow the pattern from array_cat when concatenating an empty array with another array with a null value.

select * from array_cat([], [null]);
SQLParseException[One of the arguments of the `array_cat` function can be of undefined inner type, but not both]

So it's better to allow array_append([], null)?

Reply:
> So it's better to allow array_append([], null)?
Yes I think we should tweak `array_cat` instead.

Reply:
So, how to handle this? 
1. Should I not use the ensure functions from ArrayArgumentValidators, make a new ensure function or check the validity inside the evaluate function of the implementation? 
2. Should i also change array_cat according to the choice of the 1 above?

Reply:
I'd suggest to just leave out the validator, everything else like ensuring both type are matching is already ensured by the function signature matching logic.
Right, for consistency we should change `array_cat` as well, but not sure if its worth doing now (null case maybe a bit of an edge case), I'd definitely would do this as part of this PR.

Reply:
I can make the changes by the end of the day.

Just for clarification:
I'll leave out the validator so that we can append a null value to an empty array, but not allowing to append anything to a null array?
array_append([], null) -> [null] valid
array_append(null, anyvalue) -> not valid or [anyvalue]?

For array_cat I understood that we shouldn't change it as part of this PR. 
So what do you mean by this?
> I'd definitely would do this as part of this PR

That's all I think to finalize the PR.

Reply:
Sorry mistake by me, I've meant "would **not** do this as part of this PR"

Reply:
> I'll leave out the validator so that we can append a null value to an empty array, but not allowing to append anything to a null array?

I'd leave out the validator at all and thus also allow to append anything to a null array.

Reply:
ok! pushed it :)